hydrolib/core/io/ini/util.py

```diff
@@ -341,3 +341,117 @@ def validate_coordinates(coord_name: str) -> None:
+    """
+
+    def validate_number_of_coordinates(cls, values: Dict) -> Dict:
+        """
+        Validates whether the given coordinates match in number
+        to the expected value given for numCoordinates.
+
+        Args:
+            values (Dict): Dictionary of object's validated fields.
+
+        Raises:
+            ValueError: When the number of coordinates is not specified but the coordinates are.
+            ValueError: When the number of coordinates is provided but the x-coordinates or
+                y-coordinates are not.
+            ValueError: When the number of x-coordinates or the number of y-coordinates
+                does not match the number of coordinates.
+
+        Returns:
+            Dict: Validated dictionary of input class fields.
+        """
+
+        def get_value(field_name: str) -> Any:
+            return (
+                values.get(field_name.lower(), None)
+                if not str_is_empty_or_none(field_name)
+                else None
+            )
+
+        def all_values_are_none() -> bool:
+            return (
+                number_of_coordinates is None
+                and xcoordinates is None
+                and ycoordinates is None
+            )
+
+        def coordinates_given_but_none_expected() -> bool:
+            return number_of_coordinates is None and (
+                xcoordinates is not None or ycoordinates is not None
+            )
+
+        def no_coordinates_given_while_expected(
+            number_of_coordinates: int, coordinate_field_name: str
+        ) -> bool:
+            return number_of_coordinates is not None and coordinate_field_name is None
+
+        def incorrect_number_of_coordinates_given(
+            expected_number: int, actual_number: int
+        ) -> bool:
+            return expected_number != actual_number
+
+        def validate_x_and_ycoordinate_number():
+            number_of_xcoordinates = len(xcoordinates)
+            if incorrect_number_of_coordinates_given(
+                number_of_coordinates, number_of_xcoordinates
+            ):
+                raise ValueError(
+                    f"Number of x-coordinates given ({number_of_xcoordinates}) not matching"
+                    f"the {numfield_name} value {number_of_coordinates}."
+                )
+
+            number_of_ycoordinates = len(ycoordinates)
+            if incorrect_number_of_coordinates_given(
+                number_of_coordinates, number_of_ycoordinates
+            ):
+                raise ValueError(
+                    f"Number of y-coordinates given ({number_of_ycoordinates}) not matching"
+                    f"the {numfield_name} value {number_of_coordinates}."
+                )
+
+        number_of_coordinates = get_value(numfield_name)
+        xcoordinates = get_value(xfield_name)
+        ycoordinates = get_value(yfield_name)
+
+        if all_values_are_none():
+            return values
+
+        if coordinates_given_but_none_expected():
```

Comment:
To reduce the complexicity I would consider replacing the `coordinates_given_but_none_expected()` and `no_coordinates_given_while_expected()` methods with a `some_values_are_none()` method and then using it something like:

```
        if all_values_are_none():
            return values

        if some_values_are_none():
            raise ValueError(
                f"When using coordinates, the fields {numfield_name}, {xfield_name} and {yfield_name} should be given."
            )

        validate_x_and_ycoordinate_number()
```

And the same goes for the coordinate count:
```
        def validate_x_and_ycoordinate_number():
            if len(xcoordinates) != number_of_coordinates or len(ycoordinates) != number_of_coordinates:
                raise ValueError(
                    "Number of x-coordinates and y-coordinates should match number of coordinates."
                )
```

We see that our validation methods often become very nested, because we're trying to be as specific as possible for the user. However, as discussed with @arthurvd , we would like to validate for the correct situation, instead of all the incorrect situations. Because we only have one correct situations, it is easier to validate that instead of validating all the situations that can be wrong.
I hope that makes sense and please let me know what you think.

Please also see #272

Reply:
I agree that the complexity is a bit high. I was indeed trying to be as specific as possible for the users. But if the users do not care for such specific error messages, your suggestion is indeed way more readable.

If this topic was already discussed before and the conclusion was that it is better to validate for the correct sitatuation, I will do that from now on.

We can always be more specific later if the users decide they want more specific errors!